### DISCUSSION TITLE
`RuntimeError: Tried to call pure virtual function` problems

### DISCUSSION THREAD

**Discussion by brettc** [pybind/pybind11#4013] [Q&A]
Hi, I'm not sure if this is a bug, but it was initially surprising behaviour. In any case, I'm wondering best how to work around it. I'm creating a c++ object that has a member which is a `vector<shared_ptr<T>>`, where the T is an object that implements an interface via a virtual function. I want to code some of those interfaces in C++, and others in python.

Here is a concrete example, using the Animal examples from the tutorial as a beginning. 

Full code with a working/failing test is here: https://github.com/brettc/pybind11-debugging/blob/master/

The only change is that the Animal class is now using a `shared_ptr` holder as well as specifying the trampoline class.
```c++
    py::class_<Animal, PyAnimal, std::shared_ptr<Animal>>(m, "Animal")
        .def(py::init<>())
        .def("go", &Animal::go)
        ;
    py::class_<Dog, Animal, std::shared_ptr<Dog>>(m, "Dog")
        .def(py::init<>())
        ;

```

I now create `Pound` class that contains the `vector<shared_ptr<Animal>>`

```c++
using Animal_ptr = std::shared_ptr<Animal>;
using Animal_vec = std::vector<Animal_ptr>;
using string_vec = std::vector<std::string>;

string_vec call_go_list(Animal_vec const &animals) {
    string_vec calls;
    for (const auto &p : animals) {
        calls.push_back(p->go(1));
    }
    return calls;
}

class Pound {
   public:
    Animal_vec animals_;
    explicit Pound(Animal_vec animals) : animals_(std::move(animals)) {}
    string_vec go() { return call_go_list(animals_); }
};
```

Now we can create a pound:
```python
class Cat(m.Animal):
    def go(self, n_times):
        return "meow! " * n_times

cats = [Cat() for _ in range(3)]
p = m.Pound(cats)

assert p.go() == ["meow! "] * 3
```

That works okay. But you can break it.

```python
cats = [Cat() for _ in range(3)]
p = m.Pound(cats)

# This appears to kill off *all* the Cats. Just leaving the PyAnimal class
# which can no longer locate the python objects.
del cats

# BOOM! Virtual Call!
assert p.go() == ["meow! "] * 3
```

Note that it works fine with Dogs (the C++ class). And I think I understand why this happens. But I want to know:
1. If there is some silly option I have missed that would make this work.
2. If not, how can I work around this.

**Comment by brettc** [comment]
OK, now I understand the problem better, I see that it has bitten many people. It looks like some workarounds are available. This is a good starting resource: https://github.com/pybind/pybind11/pull/2839

**Reply by leslinice** [reply]
Really

**Comment by mcleantom** [comment]
Has anyone solved this? I am getting the same issue
